src/object/object_primitive.c

```diff
@@ -16598,3 +16849,259 @@ pr_write_compressed_string_to_buffer (OR_BUF * buf, char *compressed_string, int
+ */
+
+int
+pr_clear_compressed_string (DB_VALUE * value)
+{
+  char *data = NULL;
+  DB_TYPE db_type;
+
+  if (value == NULL || DB_IS_NULL (value))
+    {
+      return NO_ERROR;		/* do nothing */
+    }
+
+  db_type = DB_VALUE_DOMAIN_TYPE (value);
+
+  /* Make sure we clear only for VARCHAR and VARNCHAR types. */
+  if (db_type != DB_TYPE_VARCHAR && db_type != DB_TYPE_VARNCHAR)
+    {
+      return NO_ERROR;		/* do nothing */
+    }
+
+  if (value->data.ch.info.compressed_need_clear == false)
+    {
+      return NO_ERROR;
+    }
+
+  if (value->data.ch.medium.compressed_size <= 0)
+    {
+      return NO_ERROR;		/* do nothing */
+    }
+
+  data = value->data.ch.medium.compressed_buf;
+  if (data != NULL)
+    {
+      db_private_free_and_init (NULL, data);
+    }
+
+  DB_SET_COMPRESSED_STRING (value, NULL, 0, false);
+
+  return NO_ERROR;
+}
+
+/* 
+ * pr_do_db_value_string_compression()	  :- Test a DB_VALUE for VARCHAR and VARNCHAR types and do string compression
+ *					  :- for such types.
+ *
+ * return()				  :- NO_ERROR or error code.
+ * value(in/out)			  :- The DB_VALUE to be tested.
+ */
+
+int
+pr_do_db_value_string_compression (DB_VALUE * value)
+{
+  DB_TYPE db_type;
+  char *compressed_string, *string;
+  int rc = NO_ERROR;
+  int src_size = 0, compressed_size;
+
+  if (value == NULL || DB_IS_NULL (value))
+    {
+      return rc;		/* do nothing */
+    }
+
+  db_type = DB_VALUE_DOMAIN_TYPE (value);
+
+  /* Make sure we clear only for VARCHAR and VARNCHAR types. */
+  if (db_type != DB_TYPE_VARCHAR && db_type != DB_TYPE_VARNCHAR)
+    {
+      return rc;		/* do nothing */
+    }
+
+  /* Make sure the value has not been through a compression before */
+  if (value->data.ch.medium.compressed_size != 0)
+    {
+      return rc;
+    }
+
+  string = db_get_string (value);
+  src_size = db_get_string_size (value);
+
+  if (src_size < PRIM_MINIMUM_STRING_LENGTH_FOR_COMPRESSION)
+    {
+      /* No compression needed. */
+      value->data.ch.medium.compressed_buf = NULL;
+      value->data.ch.medium.compressed_size = -1;
+      return rc;
+    }
+
+  /* Alloc memory for compression */
+  compressed_string = db_private_alloc (NULL, LZO_COMPRESSED_STRING_SIZE (src_size));
+  if (compressed_string == NULL)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY,
+	      1, (size_t) (LZO_COMPRESSED_STRING_SIZE (src_size)));
+      rc = ER_OUT_OF_VIRTUAL_MEMORY;
+      return rc;
+    }
+
+  rc = pr_data_compress_string (string, src_size, compressed_string, &compressed_size);
+  if (rc != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      goto cleanup;
+    }
+
+  if (compressed_size > 0)
+    {
+      /* Compression successful */
+      DB_SET_COMPRESSED_STRING (value, compressed_string, compressed_size, true);
+      return rc;
+    }
+  else
+    {
+      /* Compression failed */
+      DB_SET_COMPRESSED_STRING (value, NULL, -1, false);
+      goto cleanup;
+    }
+
+cleanup:
```

Comment:
I think it would be better to rename the goto label as "error" since it executes only in case of error (compressed _size == 0 is still error, since it cannot be a valid state)